bulk_query/query.go

```diff
@@ -597,52 +608,71 @@ func (q *QueryBenchmarker) processStats(telemetrySink chan *report.Point) {
 		}
 		// print stats to stderr (if printInterval is greater than zero):
 		if q.printInterval > 0 && i > 0 && i%q.printInterval == 0 && (int64(i) < q.limit || q.limit < 0) {
-			_, err := fmt.Fprintf(os.Stderr, "%s: after %d queries with %d workers:\n", time.Now().String(), i-q.burnIn, q.workers)
-			if err != nil {
-				log.Fatal(err)
-			}
-			fprintStats(os.Stderr, q.statMapping)
-			_, err = fmt.Fprintf(os.Stderr, "\n")
-			if err != nil {
-				log.Fatal(err)
-			}
+			log.Printf("%s: after %d queries with %d workers:\n", time.Now().String(), i-q.burnIn, q.workers)
+			fprintStats(os.Stderr, q)
+			log.Printf("\n")
 		}
 
 	}
 
-	// the final stats output goes to stdout:
-	_, err := fmt.Printf("run complete after %d queries with %d workers:\n", i-q.burnIn, q.workers)
-	if err != nil {
-		log.Fatal(err)
-	}
-	fprintStats(os.Stdout, q.statMapping)
+	log.Printf("run complete after %d queries with %d workers:\n", i-q.burnIn, q.workers)
+	q.totalQueries = int(i)
 	q.statGroup.Done()
 }
 
 // fprintStats pretty-prints stats to the given writer.
-func fprintStats(w io.Writer, statGroups StatsMap) {
+func fprintStats(w io.Writer, q *QueryBenchmarker) {
 	maxKeyLength := 0
-	keys := make([]string, 0, len(statGroups))
-	for k := range statGroups {
+	keys := make([]string, 0, len(q.statMapping))
+	for k := range q.statMapping {
 		if len(k) > maxKeyLength {
 			maxKeyLength = len(k)
 		}
 		keys = append(keys, k)
 	}
 	sort.Strings(keys)
 	for _, k := range keys {
-		v := statGroups[k]
+		v := q.statMapping[k]
 		minRate := 1e3 / v.Min
 		meanRate := 1e3 / v.Mean
 		maxRate := 1e3 / v.Max
 		paddedKey := fmt.Sprintf("%s", k)
 		for len(paddedKey) < maxKeyLength {
 			paddedKey += " "
 		}
-		_, err := fmt.Fprintf(w, "%s : min: %8.2fms (%7.2f/sec), mean: %8.2fms (%7.2f/sec), max: %7.2fms (%6.2f/sec), count: %8d, sum: %5.1fsec \n", paddedKey, v.Min, minRate, v.Mean, meanRate, v.Max, maxRate, v.Count, v.Sum/1e3)
+		kStats := make(map[string]interface{})
+		kStats["min"] = v.Min
+		kStats["minRate"] = minRate
+		kStats["mean"] = v.Mean
+		kStats["meanRate"] = meanRate
+		kStats["max"] = v.Max
+		kStats["maxRate"] = maxRate
+		kStats["count"] = v.Count
+		kStats["sum"] = v.Sum / 1e3
+		q.json[k] = kStats
+		if !q.doJson {
+			_, err := fmt.Fprintf(w, "%s : min: %8.2fms (%7.2f/sec), mean: %8.2fms (%7.2f/sec), max: %7.2fms (%6.2f/sec), count: %8d, sum: %5.1fsec \n", paddedKey, v.Min, minRate, v.Mean, meanRate, v.Max, maxRate, v.Count, v.Sum/1e3)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+	}
+	q.json["totalQueries"] = q.totalQueries
+	q.json["wallClockTime"] = q.wallTook.Seconds()
+	q.json["queryRate"] = float64(q.totalQueries) / float64(q.wallTook.Seconds())
+	q.json["workers"] = q.workers
+	q.json["batchSize"] = q.batchSize
+	if q.doJson {
+		for k, v := range q.json {
+			if _, err := json.Marshal(v); err != nil {
```

Comment:
What values of `v` would we expect to fail `json.Marshal` here?

Reply:
Some of the stats can come up as `+Inf` or `-Inf`, which Marshal cannot handle.